Color highlights for inaccuracy/mistake/blunder inside the movelist / tree view

yuzisee opened the pull request:
## Component

Movelist / Tree View

## Feature

Some basic color coding for inaccuracies vs. mistakes vs. blunders.

Follows the Lichess nomenclature of win percentage change (±0.1, ±0.2, and ±0.3) from https://github.com/lichess-org/lila/commit/133a1ebe761c4a37debc64cdaa8371e8580a1c5d

## Screenshots

![image](https://github.com/rooklift/nibbler/assets/523543/154101ea-becb-480b-a77f-9bfe5d79f670)

yuzisee commented:
FYI I've tested this out quite a bit locally, and stomped out most of the bugs. Let me know if you have any other questions!

OverLordGoldDragon commented:
Thanks for your work on this! I've read in #235 that it already works for you - so is it good to go to simply clone your branch and follow the install instructions (Windows 11)? And it supports Stockfish, correct?

yuzisee commented:
> so is it good to go to simply clone your branch and follow the install instructions

Yes.

If you already have nibbler installed elsewhere, you can also directly add the files from the git repo's `files/src` folder directly into your existing nibbler's `resources/app` folder.

Here's a zip you can download that runs on windows and has #235 #237 #238 #239 included: https://www.dropbox.com/scl/fi/qpzh38axktnof6e3clup3/nibbler-2.4.dev-windows.zip?rlkey=wzd7i1fajocd54lx1mtehd0ro

> And it supports Stockfish, correct?

I didn't touch anything Leela/Stockfish specific, so hopefully it does. Give it a try and let us know!

OverLordGoldDragon commented:
[PGN](https://github.com/rooklift/nibbler/files/12399569/lichess_pgn_2023.08.20_OverLordGoldDragon_vs_Chessarb.sgNEfoPR.zip) -- [auto example](https://github.com/rooklift/nibbler/assets/16495490/71b84c42-b69c-40e0-9f0b-11a24d077f93) -- [after passing through manually](https://github.com/rooklift/nibbler/assets/16495490/86a39a4a-0e85-41c5-9c2f-18b351238692)

yuzisee commented:
For some reason the PGN you attached is different than the one in your screenshot
![image](https://github.com/rooklift/nibbler/assets/523543/d4f62cc3-9ec9-4b12-a941-df6d36b25581)
but I'm definitely getting underlines.

What settings are you using here:
![image](https://github.com/rooklift/nibbler/assets/523543/c1ef978b-b6d7-432b-83c2-25827211730c)
and here
![image](https://github.com/rooklift/nibbler/assets/523543/3cafb047-bc2c-4351-9a4f-bd40ab0da38e)

yuzisee commented:
@OverLordGoldDragon 

Would you mind testing to see if it will work, in the meantime, by enabling **Q** here 
![image](https://github.com/rooklift/nibbler/assets/523543/9b1c57d9-90ba-4aa9-a989-0a85665268e6)


If that works, I can try updating the code so it doesn't require that setting and always appears.

OverLordGoldDragon commented:
Sorry -- [correct PGN](https://github.com/rooklift/nibbler/files/12401187/lichess_pgn_2023.08.20_MelesioG_vs_OverLordGoldDragon.tH4EMTLH.zip)

I changed all settings to yours and now it works - the one at which it worked is 1M -> 10M nodes, so maybe because it's too fast (I set to 16 cores and my CPU is very vast, it passes several moves per second). Also,

 3. `Forget all analysis` doesn't clear the annotations, I restart Nibbler each time

_(Side question: any docs on Nibbler? I can't tell what many of the options do)_

yuzisee commented:
> `Forget all analysis` doesn't clear the annotations, I restart Nibbler each time

Oh, thanks for catching that! I probably need to add a line or two to https://github.com/rooklift/nibbler/blob/004e66b2fb5b7fdc4b06da705117e40ce3e7fe6f/files/src/renderer/51_node.js#L479

Let me take a look at that.

> _(Side question: any docs on Nibbler? I can't tell what many of the options do)_

Not sure, but @rooklift I'm happy to either start some docs or refresh anything that might already exist. Happy to contribute some cleanup here and there or expand any existing documentation if helpful.

OverLordGoldDragon commented:
Well in our "quest" to mimic Lichess I think it crucial to replicate the [evaluation graph](https://github.com/rooklift/nibbler/issues/242) - @ rooklift's given me a pointer but I've not had a chance to dig into it.

I've also noticed @ rooklift's comment on another thread which I think belongs in README - i.e. if using classical engines, ensure `MultiPV=1`, and set `Threads` and `Hash` (from Stockfish forums I hear recommended is <70% of RAM).

yuzisee commented:
Discussion, continuing from https://github.com/rooklift/nibbler/issues/242#issuecomment-1694408704 and building on https://github.com/rooklift/nibbler/issues/242#issuecomment-1717694397

- If we follow the Lichess convention of "blunder" being a 25%+ change in winpct, Lichess is essentially arguing that a blunder is 2.98 centipawns from a drawn position. For Stockfish, a blunder is 1.0 centipawns from a drawn position on move 32.
- If we follow the Lichess convention of "mistake" being a 10%+ change in winpct, Lichess considers 1.1 centipaws from a drawn position to be a mistake. Stockfish would consider 0.67 centipawns to be a mistake from a drawn position on move 32.
- If we follow the Lichess convention of "inaccuracy" being a 5%+ change in winpct, any winrate past 95% can no longer produce inaccuracies for your opponent because the winpct can't increase by 5% in a single move without becoming more than 100%. Therefore:
  - An eval of **±1.65** is the point at which Stockfish evals could be considered "virtually the same" according to the Stockfish WDL model.
  - An eval of **±7.99** is the point at which Stockfish evals could be considered "virtually the same" according to Lichess live player data.


|formula|−7.99 |−2.98|−1.65|+1.65|+2.98|+7.99|
|--|--|--|--|--|--|--|
|SF v16 @ move 0| | |98.8% B : 1.2% W |  1.2% B : 98.8% W | ||
|SF v16 @ move 32| | |98.6% B : 1.4% W | 1.4% B : 98.6% W |||
|SF v16 @ move ≥120 | | |  **95% B** : 5% W | 5% B : 95% W | ||
|Lichess 2023| **95% B** | **75% B** | 64.7% B | 64.7% W | 75% W | 95% W|

## Conclusion

The suggestions in https://github.com/rooklift/nibbler/issues/242#issuecomment-1717694397 were to use 0.25 & 0.5, but they would probably be too small in practice.

|formula|eval −0.5 |−0.25|+0.25|+0.5|
|--|--|--|--|--|
|Lichess 2023| 54.6% B : 45.4% W | 52.3% B : 47.7% W |47.7% B : 52.3% W  | 45.4% B : 54.6% W|
|SF v16 @ move 0| 54.7% B : 45.3% W | 51.3% B : 48.7% W |  48.7% B : 51.3% W |  45.3% B : 54.7% W |
|SF v16 @ move 32| 53.1% B : 46.9% W | 50.8% B : 49.2% W | 49.2% B : 50.8% W | 46.9% B : 53.1% W |
|SF v16 @ move ≥120 | 52.9% B : 47.1% W  |  50.9% B : 49.1% W | 49.1% B : 50.9% W | 47.1% B : 52.9% W|

Maybe:

- ±0.5 centipawns = inaccuracy
- ±1.0 centipawns = mistake
- ±3.0 centipawns = blunder

within the stipulation that all evals larger than ±2.5 are to be considered virtually the same.

yuzisee commented:
Rebased onto latest master, and updated after https://github.com/rooklift/nibbler/pull/244

---

For posterity, here's how the other platforms handle it:

# chess.com

Note that chess.com also has a "Missed Win" annotation that is more severe than "Blunder".

# Lichess

The [Lichess rules](https://github.com/lichess-org/lila/commit/133a1ebe761c4a37debc64cdaa8371e8580a1c5d) are slightly different, and are based instead on win percentage change (±0.1, ±0.2, and ±0.3/0.5 — NOTE: Lichess defines [cpWinningChances ranging from −1.0 to +1.0](https://github.com/abebinder/lila/blob/77fe2205e30ab17ff8e028d949d2db83e41cc210/modules/analyse/src/main/Advice.scala#L57-L64)). Longer discussion is here https://github.com/rooklift/nibbler/issues/242#issuecomment-1697705967

yuzisee commented:
- Tested working
- Screenshot updated above
